File: kangaroo_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/getPublicPromotions/GetPublicPromotionsHandler.kt
@file:Suppress("INLINE_FROM_HIGHER_PLATFORM")
@file:OptIn(ExperimentalJsExport::class)
package com.kangaroo.flutterplugin.android.features.getPublicPromotions

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kotlin.js.ExperimentalJsExport
import features.getPublicPromotions.GetPublicPromotionsApi
import features.getPublicPromotions.models.PromotionsModel

import features.getPublicPromotions.serializeGetPublicPromotionsState
import kangaroorewards.appsdk.core.domain.SerializedResult
import kangaroorewards.appsdk.core.domain.toJsonResult


class GetPublicPromotionsHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/get_public_promotions"

    override val eventChannel: String
        get() = "customer_sdk/events/get_public_promotions"

    override suspend fun onMethodCall(call: MethodCall): String? {
        return getPublicPromotions(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        suspend fun getPublicPromotions(call: MethodCall): String? {
            val result = GetPublicPromotionsApi().getPublicPromotions(
                overrideHeaders = call.argument<Map<String, String>>("overrideHeaders") as Map<String, String>?,
                campaignId = call.argument<String>("campaignId") as String
            ).toJsonResult<PromotionsModel>()

            return when (result) {
                is SerializedResult.Success -> result.data
                is SerializedResult.UnauthorizedError -> result.error
                is SerializedResult.UnknownError -> result.error
                else -> null
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        GetPublicPromotionsApi().getPublicPromotionsState.serializeGetPublicPromotionsState().observe { result ->
            pushSerializedResultToEventSink<GetPublicPromotionsApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}

File: kangaroo_app_sdk_macos/macos/Classes/features/getPublicPromotions/GetPublicPromotionsHandler.swift
import Foundation
import FlutterMacOS
import KangarooAppSdkCustomer

class GetPublicPromotionsHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/get_public_promotions"

    var eventChannel: String = "customer_sdk/events/get_public_promotions"

    func onMethodCall(call: FlutterMethodCall) async -> Any? {
        return await GetPublicPromotionsHandler.getPublicPromotions(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func getPublicPromotions(call: FlutterMethodCall) async -> String? {


        


        

        guard let args = call.arguments else {
            return nil
        }
        do {
       if let myArgs = args as? [String: Any] {
          let overrideHeaders = myArgs["overrideHeaders"] as? [String: String]
                        guard let campaignId = myArgs["campaignId"] as? String else {return nil}

        let result = try await GetPublicPromotionsApi().getPublicPromotions(
                overrideHeaders: overrideHeaders,
                campaignId: campaignId
           ).serializeGetPublicPromotionsApiResult()

        switch result {
            case let result as SerializedResultSuccess:
                return result.data
            case let result as SerializedResultUnauthorizedError:
                return result.error
            case let result as SerializedResultUnknownError:
                return result.error
            default:
                return nil
                }
            }
        }
        catch {
            return nil
        }
        
        return nil
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        GetPublicPromotionsApi().getPublicPromotionsState
            .serializeGetPublicPromotionsState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultEmptyResponse:
                self.sink?(result.body)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
